File: src/tokamak/extract.py
"""Locate reasoning spans inside heterogeneous trace formats.

Supports:
- Hermes-style JSONL — top-level `messages` list
- Claude/Anthropic message JSONL — `messages` list with `content` blocks
- OpenAI o1-style — message has a `reasoning` field
- Inline tag form — `<thinking>...</thinking>` etc. anywhere in any string

A reasoning span is described as an editable accessor: a (getter, setter)
pair so the patcher can rewrite in place without knowing the trace shape.

Each span optionally carries the surrounding PROBLEM and ANSWER context
extracted from the same message list, so downstream agents can be fed the
full rail (problem + answer + reasoning) instead of the reasoning alone.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple

REASONING_TAGS = ["thinking", "think", "reasoning", "thought", "analyze", "scratchpad"]

_TAG_RE = re.compile(
    r"<(" + "|".join(REASONING_TAGS) + r")>(.*?)</\1>",
    re.DOTALL | re.IGNORECASE,
)


@dataclass
class Span:
    source: str                                # "tag:thinking" / "field:reasoning" / "block:thinking"
    text: str                                  # current contents of the reasoning span
    apply: Callable[[str], None]               # write replacement back into the trace
    problem: str = ""                          # surrounding user / system context
    answer: str = ""                           # the visible assistant answer that follows


def _walk(obj: Any, path: List[Any]) -> Iterator[Tuple[List[Any], Any]]:
    yield path, obj
    if isinstance(obj, dict):
        for k, v in obj.items():
            yield from _walk(v, path + [k])
    elif isinstance(obj, list):
        for i, v in enumerate(obj):
            yield from _walk(v, path + [i])


def _setter(root: Any, path: List[Any]) -> Callable[[Any], None]:
    def set_(value: Any) -> None:
        cursor = root
        for key in path[:-1]:
            cursor = cursor[key]
        cursor[path[-1]] = value
    return set_


def _replace_tag_spans(string_value: str, on_replace: Callable[[str], None]) -> List[Span]:
    """Build spans for every <thinking>...</thinking> match inside a string.

    The setter closes over a mutable buffer so multiple spans in the same
    string each rewrite their own match without clobbering siblings.
    """
    matches = list(_TAG_RE.finditer(string_value))
    if not matches:
        return []

    buf = [string_value]   # single-element list so closures share state

    spans: List[Span] = []
    for m in matches:
        tag = m.group(1)
        inner = m.group(2)
        outer_orig = m.group(0)

        def make(tag=tag, outer_orig=outer_orig):
            def apply(new_inner: str) -> None:
                replacement = f"<{tag}>{new_inner}</{tag}>"
                buf[0] = buf[0].replace(outer_orig, replacement, 1)
                on_replace(buf[0])
            return apply

        spans.append(Span(source=f"tag:{tag.lower()}", text=inner, apply=make()))
    return spans


# ---------------------------------------------------------------------------
# Context extraction — problem + answer pair surrounding a reasoning span
# ---------------------------------------------------------------------------

def _strip_reasoning_tags(text: str) -> str:
    """Remove all reasoning tag spans from a string to recover the visible answer."""
    return _TAG_RE.sub("", text).strip()


def _gather_problem(messages: List[Dict[str, Any]], assistant_idx: int) -> str:
    """Collect user/system messages preceding `assistant_idx` into a single problem string."""
    parts: List[str] = []
    for m in messages[:assistant_idx]:
        if not isinstance(m, dict):
            continue
        role = m.get("role")
        content = m.get("content")
        if not isinstance(content, str):
            # Anthropic content-block form: join text blocks.
            if isinstance(content, list):
                content = " ".join(
                    b.get("text", "") for b in content
                    if isinstance(b, dict) and b.get("type") == "text"
                )
            else:
                continue
        if role in ("system", "user", "human"):
            parts.append(content.strip())
    return "\n\n".join(p for p in parts if p)


def _extract_answer(message: Dict[str, Any]) -> str:
    """Pull the visible (non-reasoning) text out of an assistant message."""
    content = message.get("content")
    if isinstance(content, str):
        return _strip_reasoning_tags(content)
    if isinstance(content, list):
        # Anthropic-style content blocks: join the text blocks (drop thinking blocks).
        chunks: List[str] = []
        for b in content:
            if not isinstance(b, dict):
                continue
            if b.get("type") == "text" and isinstance(b.get("text"), str):
                chunks.append(_strip_reasoning_tags(b["text"]))
        return "\n\n".join(c for c in chunks if c).strip()
    return ""


def _attach_context(
    span: Span, messages: List[Dict[str, Any]], assistant_idx: int
) -> None:
    span.problem = _gather_problem(messages, assistant_idx)
    span.answer = _extract_answer(messages[assistant_idx])


def find_reasoning_spans(trace: Dict[str, Any]) -> List[Span]:
    """Walk a trace and yield every editable reasoning span we can find.

    Spans returned by this function include `.problem` and `.answer`
    populated from the surrounding message list when one is available.
    """
    spans: List[Span] = []

    messages = trace.get("messages") if isinstance(trace, dict) else None

    # Pass 1: scan messages directly so we can attach problem/answer context.
    if isinstance(messages, list):
        for ai, msg in enumerate(messages):
            if not isinstance(msg, dict):
                continue
            if msg.get("role") not in ("assistant", "model"):
                continue
            content = msg.get("content")

            if isinstance(content, str) and _TAG_RE.search(content):
                set_at_path = _setter(trace, ["messages", ai, "content"])
                new_spans = _replace_tag_spans(content, set_at_path)
                for s in new_spans:
                    _attach_context(s, messages, ai)
                spans.extend(new_spans)
                continue

            if isinstance(content, list):
                for bi, block in enumerate(content):
                    if not isinstance(block, dict):
                        continue
                    # Anthropic content-block thinking
                    if block.get("type") == "thinking" and isinstance(block.get("thinking"), str):
                        set_block = _setter(trace, ["messages", ai, "content", bi, "thinking"])
                        s = Span(
                            source="block:thinking",
                            text=block["thinking"],
                            apply=set_block,
                        )
                        _attach_context(s, messages, ai)
                        spans.append(s)
                    # OpenAI-style "reasoning" field on a content block
                    for fld in ("reasoning", "thought", "analysis"):
                        val = block.get(fld)
                        if isinstance(val, str) and val.strip():
                            set_field = _setter(trace, ["messages", ai, "content", bi, fld])
                            s = Span(
                                source=f"field:{fld}",
                                text=val,
                                apply=set_field,
                            )
                            _attach_context(s, messages, ai)
                            spans.append(s)

            # OpenAI o1: top-level reasoning field on the message itself.
            for fld in ("reasoning", "thinking", "analysis", "thought"):
                val = msg.get(fld)
                if isinstance(val, str) and val.strip():
                    set_field = _setter(trace, ["messages", ai, fld])
                    s = Span(
                        source=f"field:{fld}",
                        text=val,
                        apply=set_field,
                    )
                    _attach_context(s, messages, ai)
                    spans.append(s)

    # Pass 2: catch any reasoning tags that live outside the messages array.
    if not spans:
        for path, node in _walk(trace, []):
            if isinstance(node, str) and "<" in node and _TAG_RE.search(node):
                set_at_path = _setter(trace, path)
                spans.extend(_replace_tag_spans(node, set_at_path))

    # Dedup.
    seen = set()
    unique: List[Span] = []
    for s in spans:
        key = (s.source, id(s.apply))
        if key in seen:
            continue
        seen.add(key)
        unique.append(s)
    return unique

File: rust/tokamak/src/extract.rs
//! Locate reasoning spans across heterogeneous trace formats and capture
//! the surrounding **problem + answer** context the agents need as rails.

use once_cell::sync::Lazy;
use regex::Regex;
use serde_json::{Map, Value};

/// Each pattern matches a single tag pair. The closing tag is duplicated
/// inline since Rust's `regex` crate does not support backreferences.
static TAG_RES: Lazy<Vec<(&'static str, Regex)>> = Lazy::new(|| {
    let tags = ["thinking", "think", "reasoning", "thought", "analyze", "scratchpad"];
    tags.into_iter()
        .map(|t| {
            let pat = format!(r"(?is)<{tag}>(.*?)</{tag}>", tag = t);
            (t, Regex::new(&pat).unwrap())
        })
        .collect()
});

/// A reasoning span located in a trace, with handles to rewrite it back.
///
/// We don't carry closures — instead we record the JSON pointer-style path
/// to the host string and the original outer text to splice over. The
/// pipeline calls `apply()` to perform the rewrite.
#[derive(Debug, Clone)]
pub struct Span {
    pub source: String,        // "tag:thinking" / "block:thinking" / "field:reasoning"
    pub text: String,          // current reasoning text
    pub problem: String,
    pub answer: String,

    // Where to write the result back:
    pub path: Vec<PathStep>,   // path from trace root to the host string field
    pub kind: SpanKind,
}

#[derive(Debug, Clone)]
pub enum PathStep { Key(String), Idx(usize) }

#[derive(Debug, Clone)]
pub enum SpanKind {
    /// The reasoning lives inside `<tag>...</tag>` of the host string;
    /// rewrite by replacing the outer tag-and-body.
    TagBody { tag: String, outer_original: String },
    /// The host field is itself the reasoning text; rewrite by replacing it.
    Field,
}

/// Walk a trace and yield every editable reasoning span we can find.
pub fn find_reasoning_spans(trace: &Value) -> Vec<Span> {
    let mut out: Vec<Span> = Vec::new();
    let Value::Object(root) = trace else { return out; };
    let Some(Value::Array(messages)) = root.get("messages") else {
        return walk_fallback(trace);
    };

    for (ai, msg) in messages.iter().enumerate() {
        let Value::Object(m) = msg else { continue; };
        let role = m.get("role").and_then(|v| v.as_str()).unwrap_or("");
        if !matches!(role, "assistant" | "model") { continue; }

        let problem = gather_problem(messages, ai);
        let answer  = extract_answer(m);

        // Case A: content is a tagged string.
        if let Some(Value::String(s)) = m.get("content") {
            for (tag, re) in TAG_RES.iter() {
                for cap in re.captures_iter(s) {
                    let outer = cap.get(0).unwrap().as_str().to_string();
                    let inner = cap.get(1).unwrap().as_str().to_string();
                    out.push(Span {
                        source:  format!("tag:{tag}"),
                        text:    inner,
                        problem: problem.clone(),
                        answer:  answer.clone(),
                        path: vec![
                            PathStep::Key("messages".into()),
                            PathStep::Idx(ai),
                            PathStep::Key("content".into()),
                        ],
                        kind: SpanKind::TagBody {
                            tag: (*tag).to_string(),
                            outer_original: outer,
                        },
                    });
                }
            }
            continue;
        }

        // Case B: content is a content-block list.
        if let Some(Value::Array(blocks)) = m.get("content") {
            for (bi, b) in blocks.iter().enumerate() {
                let Value::Object(block) = b else { continue; };

                // Anthropic-style thinking block
                if block.get("type").and_then(|v| v.as_str()) == Some("thinking") {
                    if let Some(Value::String(thinking)) = block.get("thinking") {
                        out.push(Span {
                            source: "block:thinking".into(),
                            text: thinking.clone(),
                            problem: problem.clone(),
                            answer:  answer.clone(),
                            path: vec![
                                PathStep::Key("messages".into()),
                                PathStep::Idx(ai),
                                PathStep::Key("content".into()),
                                PathStep::Idx(bi),
                                PathStep::Key("thinking".into()),
                            ],
                            kind: SpanKind::Field,
                        });
                    }
                }

                for fld in ["reasoning", "thought", "analysis"] {
                    if let Some(Value::String(val)) = block.get(fld) {
                        if !val.trim().is_empty() {
                            out.push(Span {
                                source: format!("field:{fld}"),
                                text: val.clone(),
                                problem: problem.clone(),
                                answer:  answer.clone(),
                                path: vec![
                                    PathStep::Key("messages".into()),
                                    PathStep::Idx(ai),
                                    PathStep::Key("content".into()),
                                    PathStep::Idx(bi),
                                    PathStep::Key(fld.into()),
                                ],
                                kind: SpanKind::Field,
                            });
                        }
                    }
                }
            }
        }

        // Case C: OpenAI o1-style `reasoning` field on the message itself.
        for fld in ["reasoning", "thinking", "analysis", "thought"] {
            if let Some(Value::String(val)) = m.get(fld) {
                if !val.trim().is_empty() {
                    out.push(Span {
                        source: format!("field:{fld}"),
                        text: val.clone(),
                        problem: problem.clone(),
                        answer:  answer.clone(),
                        path: vec![
                            PathStep::Key("messages".into()),
                            PathStep::Idx(ai),
                            PathStep::Key(fld.into()),
                        ],
                        kind: SpanKind::Field,
                    });
                }
            }
        }
    }

    out
}

/// Fallback walker for traces without a top-level `messages` array.
fn walk_fallback(trace: &Value) -> Vec<Span> {
    let mut out = Vec::new();
    walk(trace, &mut Vec::new(), &mut out);
    out
}

fn walk(node: &Value, path: &mut Vec<PathStep>, out: &mut Vec<Span>) {
    match node {
        Value::String(s) => {
            for (tag, re) in TAG_RES.iter() {
                for cap in re.captures_iter(s) {
                    let outer = cap.get(0).unwrap().as_str().to_string();
                    let inner = cap.get(1).unwrap().as_str().to_string();
                    out.push(Span {
                        source: format!("tag:{tag}"),
                        text: inner,
                        problem: String::new(),
                        answer:  String::new(),
                        path: path.clone(),
                        kind: SpanKind::TagBody {
                            tag: (*tag).to_string(),
                            outer_original: outer,
                        },
                    });
                }
            }
        }
        Value::Object(obj) => {
            for (k, v) in obj.iter() {
                path.push(PathStep::Key(k.clone()));
                walk(v, path, out);
                path.pop();
            }
        }
        Value::Array(arr) => {
            for (i, v) in arr.iter().enumerate() {
                path.push(PathStep::Idx(i));
                walk(v, path, out);
                path.pop();
            }
        }
        _ => {}
    }
}

// ---- Context helpers ----

fn gather_problem(messages: &[Value], assistant_idx: usize) -> String {
    let mut parts = Vec::new();
    for m in &messages[..assistant_idx] {
        let Value::Object(o) = m else { continue; };
        let role = o.get("role").and_then(|v| v.as_str()).unwrap_or("");
        if !matches!(role, "user" | "system" | "human") { continue; }
        let txt = content_string(o);
        if !txt.trim().is_empty() { parts.push(txt.trim().to_string()); }
    }
    parts.join("\n\n")
}

fn extract_answer(m: &Map<String, Value>) -> String {
    match m.get("content") {
        Some(Value::String(s)) => strip_reasoning(s).trim().to_string(),
        Some(Value::Array(blocks)) => {
            let chunks: Vec<String> = blocks.iter().filter_map(|b| {
                let Value::Object(block) = b else { return None; };
                if block.get("type").and_then(|v| v.as_str()) != Some("text") {
                    return None;
                }
                block.get("text").and_then(|v| v.as_str())
                    .map(|t| strip_reasoning(t))
            }).collect();
            chunks.join("\n\n").trim().to_string()
        }
        _ => String::new(),
    }
}

fn content_string(m: &Map<String, Value>) -> String {
    match m.get("content") {
        Some(Value::String(s)) => s.clone(),
        Some(Value::Array(blocks)) => {
            blocks.iter().filter_map(|b| {
                let Value::Object(block) = b else { return None; };
                block.get("text").and_then(|v| v.as_str()).map(|s| s.to_string())
            }).collect::<Vec<_>>().join("\n")
        }
        _ => String::new(),
    }
}

fn strip_reasoning(text: &str) -> String {
    let mut out = text.to_string();
    for (_tag, re) in TAG_RES.iter() {
        out = re.replace_all(&out, "").into_owned();
    }
    out
}

// ---- Apply replacement back into the trace ----

pub fn apply_replacement(trace: &mut Value, span: &Span, new_inner: &str) {
    // Walk to the parent value.
    let mut cursor: &mut Value = trace;
    for step in &span.path[..span.path.len().saturating_sub(1)] {
        cursor = match (step, cursor) {
            (PathStep::Key(k), Value::Object(o)) => o.get_mut(k).unwrap(),
            (PathStep::Idx(i), Value::Array(a))  => a.get_mut(*i).unwrap(),
            _ => return,
        };
    }
    let Some(last) = span.path.last() else { return; };
    let target: &mut Value = match (last, cursor) {
        (PathStep::Key(k), Value::Object(o)) => o.get_mut(k).unwrap(),
        (PathStep::Idx(i), Value::Array(a))  => a.get_mut(*i).unwrap(),
        _ => return,
    };

    match &span.kind {
        SpanKind::Field => {
            *target = Value::String(new_inner.to_string());
        }
        SpanKind::TagBody { tag, outer_original } => {
            if let Value::String(s) = target {
                let replacement = format!("<{tag}>{new_inner}</{tag}>");
                *s = s.replacen(outer_original, &replacement, 1);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn finds_thinking_tag() {
        let t = json!({"messages": [
            {"role": "user", "content": "Hi"},
            {"role": "assistant", "content": "<thinking>internal step</thinking>\nAnswer."},
        ]});
        let spans = find_reasoning_spans(&t);
        assert_eq!(spans.len(), 1);
        assert_eq!(spans[0].source, "tag:thinking");
        assert!(spans[0].text.contains("internal step"));
        assert!(spans[0].problem.contains("Hi"));
        assert!(spans[0].answer.contains("Answer."));
    }

    #[test]
    fn apply_rewrites_in_place() {
        let mut t = json!({"messages": [
            {"role": "assistant", "content": "<thinking>orig</thinking>\nA"}
        ]});
        let spans = find_reasoning_spans(&t);
        apply_replacement(&mut t, &spans[0], "compressed");
        let s = t["messages"][0]["content"].as_str().unwrap();
        assert!(s.contains("<thinking>compressed</thinking>"));
    }

    #[test]
    fn anthropic_thinking_block() {
        let t = json!({"messages":[{"role":"assistant","content":[
            {"type":"thinking","thinking":"hidden chain"},
            {"type":"text","text":"visible answer"},
        ]}]});
        let spans = find_reasoning_spans(&t);
        assert!(spans.iter().any(|s| s.source == "block:thinking"));
    }
}
